ls1intum/Artemis
# Add Option to Disable More Feedback Requests

### Checklist
- [x] The database changes got approved. --- **Do not test this PR on the test servers until this is checked!**

### Motivation and Context
Some courses need to be able to adjust the assessment after a complaint if it is wrong because exam-rules apply, practical courses such as PGdP have this requirement. Student using the more feedback function only make everything more complicated, 
because the score cannot be modified by tutors. We want to disable more feedback requests for the course and only allow (unlimited) complaints because we more or less have to.

### Description
- **Database Change: added `max_request_more_feedback_time_days` column**
  It is not nullable and a primitive int.
- Added a corresponding attribute to the `Course` class and adjusted some logic
- Changed course and complaint validation (on create/update)
- Added the attribute on the client side as well
- Adjust client forms and some complaint-setting-based conditions

#### Open Questions
- [ ] Find out why team complaints do not work at all (even on develop, or am I doing something wrong?)
  Seems to be decided by `this.isCurrentUserSubmissionAuthor = this.participation.student.id === user.id;`, is that a problem?
- [ ] The same goes for more feedback requests for teams, those don't seem to exist as well.

### Steps for Testing
0. Review the code, especially look if I missed something that needs to be changed as well
1. Test that complaints with various course settings work:
    1. Request more feedback enabled <-> disabled
    4. Complaints enabled <-> disabled.
2. Adjust the course setting and then see, if the correct options are available for an _already graded submission in this course_.
    "Request More Feedback" should only be available if the option is checked, and "Complain" only when complaint tokens are available.
3. Submit a Complaint / More Feedback Request and check that it is accepted by the server (e.g. check on the Instructor Dashboard)

### Test Coverage
_more or less unchanged_

### Screenshots
- **Options**
  ![image](https://user-images.githubusercontent.com/11130248/99712778-cf3f4c80-2aa3-11eb-8018-d8c789529ab6.png)
- **Students can only submit complaints and not request more feedback if disabled**
  ![image](https://user-images.githubusercontent.com/11130248/96325931-c682d300-102c-11eb-9998-6d8b4ea22525.png)
